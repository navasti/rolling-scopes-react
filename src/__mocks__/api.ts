import {
  PokemonMoveDetails,
  PokemonTypeDetails,
  PokemonDetails,
  PokemonMove,
  PokemonType,
  Pokemon,
} from 'types';
import { hasError } from 'utils';

export const fetchBaseMock = (mockData: Pokemon | PokemonMove | PokemonType) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve({ results: [mockData] }),
    })
  ) as jest.Mock;
};

export const fetchByParameterMock = (
  mockData: PokemonDetails | PokemonMoveDetails | PokemonTypeDetails
) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve(mockData),
    })
  ) as jest.Mock;
};

export const fetchDetailsMock = <T>(mockData: Array<T>) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve(mockData[0]),
    })
  ) as jest.Mock;
};

export const handleResponseMock = (status: number, mockData: Pokemon) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status,
      json: () => Promise.resolve(hasError(status) ? undefined : mockData),
    })
  ) as jest.Mock;
};

export const handleMappedResponseMock = (status: number, mockData: PokemonDetails) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status,
      json: () => Promise.resolve(hasError(status) ? undefined : mockData),
    })
  ) as jest.Mock;
};
