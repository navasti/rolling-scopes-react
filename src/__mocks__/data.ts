import { ErrorMessages } from 'appConstants';
import {
  PokemonMoveDetails,
  PokemonTypeDetails,
  PokemonDetails,
  CustomPokemon,
  ErrorsObject,
  PokemonData,
  PokemonMove,
  PokemonType,
  Details,
  Pokemon,
  Lengths,
} from 'types';

export const visible = true;
export const invisible = false;
export const isLoading = false;

export const links = ['Home', 'About', 'Form', 'Invalid'];

export const testingComponentName = 'Test';
export const testingLocation = '/test';

export const testingContent = 'Content';
export const testingTitle = 'Title';

export const pokemonType: PokemonType = {
  name: 'typex',
  url: 'url',
};

export const pokemon: Pokemon = {
  name: 'pokemonix',
  url: 'url',
};

export const pokemonMove: PokemonMove = {
  name: 'movix',
  url: 'url',
};

export const typeModalContentProperties = [
  'Double damage from',
  'Double damage to',
  'Half damage from',
  'Half damage to',
  'No damage from',
  'No damage to',
  'Pokemons',
  'Moves',
] as const;

export const moveModalContentProperties = [
  'Learned by',
  'Priority',
  'Accuracy',
  'Power',
  'Type',
  'PP',
] as const;

export const pokemonModalContentProperties = [
  'Pokemon statistics',
  'Pokemon abilities',
  'Base experience',
  'Height',
  'Weight',
  'Types',
] as const;

export const pokemonData: PokemonData = {
  results: [{ name: 'testName', url: 'testURL' }],
  previous: null,
  next: null,
  count: 1,
};

export const testEmptyErrors: ErrorsObject = {
  birthday: null,
  consent: null,
  gender: null,
  name: null,
  type: null,
};

export const testErrors: ErrorsObject = {
  birthday: ErrorMessages.birthday,
  consent: ErrorMessages.consent,
  gender: ErrorMessages.gender,
  name: ErrorMessages.name,
  type: ErrorMessages.type,
};

export const customPokemon: CustomPokemon = {
  birthday: '01-01-1999',
  name: 'pokemonix',
  gender: 'male',
  avatar: null,
  type: 'fire',
  shiny: true,
};

//

export const lengths: Lengths = { pokemons: 10, moves: 15, types: 20 };

const samplePokemon = (): PokemonDetails => ({
  base_experience: 62,
  name: 'charmander',
  weight: 85,
  height: 6,
  id: 4,
  abilities: [
    {
      ability: { name: 'blaze', url: 'https://pokeapi.co/api/v2/ability/66/' },
      is_hidden: false,
      slot: 1,
    },
  ],
  sprites: {
    front_default: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/4.png',
    back_default:
      'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/back/4.png',
  },
  stats: [
    {
      base_stat: 52,
      effort: 0,
      stat: {
        name: 'attack',
        url: 'https://pokeapi.co/api/v2/stat/2/',
      },
    },
  ],
  types: [
    {
      slot: 1,
      type: {
        name: 'fire',
        url: 'https://pokeapi.co/api/v2/type/10/',
      },
    },
  ],
});

const sampleMove = (): PokemonMoveDetails => ({
  accuracy: 100,
  name: 'tackle',
  priority: 0,
  power: 40,
  id: 33,
  pp: 35,
  learned_by_pokemon: [
    {
      name: 'bulbasaur',
      url: 'https://pokeapi.co/api/v2/pokemon/1/',
    },
  ],
  type: {
    name: 'normal',
    url: 'https://pokeapi.co/api/v2/type/1/',
  },
});

const sampleType = (): PokemonTypeDetails => ({
  moves: [{ name: 'gust', url: 'https://pokeapi.co/api/v2/move/16/' }],
  name: 'flying',
  id: 3,
  damage_relations: {
    double_damage_from: [
      { name: 'electric', url: 'https://pokeapi.co/api/v2/type/13/' },
      { name: 'rock', url: 'https://pokeapi.co/api/v2/type/6/' },
    ],
    double_damage_to: [{ name: 'fighting', url: 'https://pokeapi.co/api/v2/type/2/' }],
    half_damage_from: [{ name: 'bug', url: 'https://pokeapi.co/api/v2/type/7/' }],
    half_damage_to: [{ name: 'steel', url: 'https://pokeapi.co/api/v2/type/9/' }],
    no_damage_from: [{ name: 'ground', url: 'https://pokeapi.co/api/v2/type/5/' }],
    no_damage_to: [],
  },
  pokemon: [
    { pokemon: { name: 'charizard', url: 'https://pokeapi.co/api/v2/pokemon/6/' }, slot: 2 },
  ],
});

const prepareBuilder = <T extends Details>(sample: () => T) => {
  const builder = (object = sample()) => ({
    withName: (name: T['name']) => builder({ ...object, name }),
    withId: (id: T['id']) => builder({ ...object, id }),
    valueOf: () => object,
  });
  return builder;
};

const pokemonBuilder = prepareBuilder(samplePokemon);
const moveBuilder = prepareBuilder(sampleMove);
const typeBuilder = prepareBuilder(sampleType);

export const pokemonsMock: Array<PokemonDetails> = [
  pokemonBuilder().withId(1).withName('bulbasaur'),
  pokemonBuilder().withId(25).withName('pikachu'),
  pokemonBuilder(),
].map((builder) => builder.valueOf());

export const movesMock: Array<PokemonMoveDetails> = [
  moveBuilder().withId(55).withName('water-gun'),
  moveBuilder().withId(50).withName('disable'),
  moveBuilder(),
].map((builder) => builder.valueOf());

export const typesMock: Array<PokemonTypeDetails> = [
  typeBuilder().withId(2).withName('fighting'),
  typeBuilder().withId(5).withName('ground'),
  typeBuilder(),
].map((builder) => builder.valueOf());
