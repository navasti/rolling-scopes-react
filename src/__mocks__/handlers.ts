import { screen } from '@testing-library/react';
import { AvailableTabs } from 'appConstants';

export const expectByText = jest.fn((text: string) => {
  expect(screen.getByText(text)).toBeInTheDocument();
});

export const handleCloseModal = jest.fn();
export const handleCardSelect = jest.fn();
export const handleOpenModal = jest.fn();
export const onTabClick = jest.fn((availableTab: AvailableTabs) => availableTab);
