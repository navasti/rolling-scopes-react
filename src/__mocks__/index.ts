export { testOnChange, handleCloseModal } from './handlers';
export { localStorageMock } from './localStorage';

export {
  fetchBaseMock,
  fetchDetailsMock,
  handleResponseMock,
  fetchByParameterMock,
  handleMappedResponseMock,
} from './api';

export {
  links,
  pokemon,
  visible,
  invisible,
  isLoading,
  movesMock,
  typesMock,
  testErrors,
  pokemonData,
  pokemonMove,
  pokemonType,
  testingTitle,
  pokemonsMock,
  customPokemon,
  testingContent,
  testingLocation,
  testEmptyErrors,
  testingComponentName,
  moveModalContentProperties,
  typeModalContentProperties,
  pokemonModalContentProperties,
} from './data';

export {
  modalRef,
  errorRef,
  inputRef,
  selectRef,
  ModalTitle,
  ModalContent,
  TestingElement,
  TestingErrorElement,
} from './elements';
