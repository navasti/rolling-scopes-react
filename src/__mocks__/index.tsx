import { ErrorMessages } from 'appConstants';
import { createRef, RefObject } from 'react';
import { hasError } from 'utils';
import {
  MockDetailedDataType,
  PokemonMoveDetails,
  PokemonTypeDetails,
  MockBaseDataType,
  PokemonDetails,
  CustomPokemon,
  ErrorsObject,
  CheckFields,
  PokemonData,
  PokemonMove,
  PokemonType,
  TextFields,
  Pokemon,
} from 'types';

export const visible = true;
export const invisible = false;
export const isLoading = false;

export const errorRef: RefObject<HTMLSpanElement> = createRef();
export const inputRef: RefObject<HTMLInputElement> = createRef();
export const selectRef: RefObject<HTMLSelectElement> = createRef();

export const links = ['Home', 'About', 'Form', 'Invalid'];
export const componentName = 'About';
export const location = '/about';
export const testingContent = 'Content';
export const testingTitle = 'Title';

export const TestingElement = () => <p>testing element</p>;
export const TestingErrorElement = () => <p>testing error</p>;

export const TestingTitleElement = () => <span>{testingTitle}</span>;
export const TestingContentElement = () => <div>{testingContent}</div>;
export const modalRef: RefObject<HTMLDivElement> = createRef();

export const fetchBaseMock = <T extends MockBaseDataType>(mockData: T) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve({ results: [mockData] }),
    })
  ) as jest.Mock;
};

export const fetchByParameterMock = <T extends MockDetailedDataType>(mockData: T) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve(mockData),
    })
  ) as jest.Mock;
};

export const fetchDetailsMock = <T,>(mockData: Array<T>) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve(mockData[0]),
    })
  ) as jest.Mock;
};

export const pokemonType: PokemonType = {
  name: 'typex',
  url: 'url',
};

export const pokemon: Pokemon = {
  name: 'pokemonix',
  url: 'url',
};

export const pokemonMove: PokemonMove = {
  name: 'movix',
  url: 'url',
};

export const typeModalContentProperties = [
  'Double damage from',
  'Double damage to',
  'Half damage from',
  'Half damage to',
  'No damage from',
  'No damage to',
  'Pokemons',
  'Moves',
] as const;

export const moveModalContentProperties = [
  'Learned by',
  'Priority',
  'Accuracy',
  'Power',
  'Type',
  'PP',
] as const;

export const pokemonModalContentProperties = [
  'Pokemon statistics',
  'Pokemon abilities',
  'Base experience',
  'Height',
  'Weight',
  'Types',
] as const;

export const pokemonData: PokemonData = {
  results: [{ name: 'testName', url: 'testURL' }],
  previous: null,
  next: null,
  count: 1,
};

export const testOnChange = (field: TextFields | CheckFields) => console.log(field);

export const testEmptyErrors: ErrorsObject = {
  birthday: null,
  consent: null,
  gender: null,
  name: null,
  type: null,
};

export const testErrors: ErrorsObject = {
  birthday: ErrorMessages.birthday,
  consent: ErrorMessages.consent,
  gender: ErrorMessages.gender,
  name: ErrorMessages.name,
  type: ErrorMessages.type,
};

export const customPokemon: CustomPokemon = {
  birthday: '01-01-1999',
  name: 'pokemonix',
  gender: 'male',
  avatar: null,
  type: 'fire',
  shiny: true,
};

export const detailedTypes: Array<PokemonTypeDetails> = [
  {
    damage_relations: {
      double_damage_from: [],
      double_damage_to: [{ name: 'toTest', url: 'urlTest' }],
      half_damage_from: [{ name: 'fromTest', url: 'urlTest' }],
      half_damage_to: [],
      no_damage_from: [],
      no_damage_to: [],
    },
    pokemon: [{ pokemon: { name: 'pokemonTest', url: 'urlTest' }, slot: 1 }],
    moves: [{ name: 'moveTest', url: 'urlTest' }],
    name: 'typeTest',
    id: 10,
  },
];

export const detailedMoves: Array<PokemonMoveDetails> = [
  {
    learned_by_pokemon: [{ name: 'pokomonTest', url: 'urlTest' }],
    type: { name: 'typeTest', url: 'urlTest' },
    name: 'moveTest',
    accuracy: 60,
    priority: 0,
    power: 50,
    id: 20,
    pp: 15,
  },
];

export const detailedPokemons: Array<PokemonDetails> = [
  {
    abilities: [{ ability: { name: 'ability' }, is_hidden: false, slot: 2 }],
    stats: [{ base_stat: 33, effort: 33, stat: { name: 'test' } }],
    sprites: { back_default: 'url', front_default: 'url' },
    types: [{ slot: 2, type: { name: 'typeTest' } }],
    name: 'Testing pokemon',
    base_experience: 12,
    height: 15,
    weight: 10,
    id: 34,
  },
  {
    abilities: [{ ability: { name: 'ability2' }, is_hidden: true, slot: 1 }],
    stats: [{ base_stat: 22, effort: 22, stat: { name: 'test' } }],
    sprites: { back_default: 'url2', front_default: 'url2' },
    types: [{ slot: 2, type: { name: 'typeTest2' } }],
    name: 'Testing pokemon2',
    base_experience: 16,
    height: 11,
    weight: 11,
    id: 35,
  },
];

type Store = Record<string, string>;

export const localStorageMock = (() => {
  let store: Store = {};
  return {
    getItem(key: string) {
      return store[key];
    },
    setItem(key: string, value: string) {
      store[key] = value.toString();
    },
    clear() {
      store = {};
    },
    removeItem(key: string) {
      delete store[key];
    },
  };
})();
