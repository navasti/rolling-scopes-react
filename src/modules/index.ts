export { LocationBadge } from './Header/components';
export { DetailsModal } from './DetailsModal';
export { SearchPage } from './SearchPage';
export { NotFound } from './NotFound';
export { Header } from './Header';
export { Layout } from './Layout';
export { About } from './About';
export { Form } from './Form';
export { App } from './App';
