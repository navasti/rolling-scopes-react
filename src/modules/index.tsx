export { App } from './App';
export { Form } from './Form';
export { About } from './About';
export { Layout } from './Layout';
export { Header } from './Header';
export { NotFound } from './NotFound';
export { SearchPage } from './SearchPage';
export { LocationBadge } from './Header/components';
export { Card, SearchBar } from './SearchPage/components';
