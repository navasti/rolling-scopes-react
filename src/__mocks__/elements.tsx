import { createRef, RefObject } from 'react';
import { testingContent, testingTitle } from './data';

export const modalRef: RefObject<HTMLDivElement> = createRef();
export const errorRef: RefObject<HTMLSpanElement> = createRef();
export const inputRef: RefObject<HTMLInputElement> = createRef();
export const selectRef: RefObject<HTMLSelectElement> = createRef();

export const TestingElement = () => <p>testing element</p>;
export const TestingErrorElement = () => <p>testing error</p>;

export const ModalTitle = () => <span data-testid="modal-title">{testingTitle}</span>;
export const ModalContent = () => <div data-testid="modal-content">{testingContent}</div>;
