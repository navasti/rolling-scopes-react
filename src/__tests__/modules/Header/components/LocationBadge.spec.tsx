import { testingLocation, testingComponentName } from '__mocks__/data';
import { render, screen } from '@testing-library/react';
import { LocationBadge } from 'modules';

describe('LocationBadge', () => {
  it('location badge element should be rendered and match snapshot', () => {
    const { container } = render(
      <LocationBadge componentName={testingComponentName} location={testingLocation} />
    );
    expect(container).toMatchSnapshot();
    expect(container.children[0]).toBeInTheDocument();
  });
  it('location and component name should be rendered and visible', () => {
    render(<LocationBadge componentName={testingComponentName} location={testingLocation} />);
    expect(screen.getByText(testingComponentName)).toBeVisible();
    expect(screen.getByText(testingLocation)).toBeVisible();
  });
});
