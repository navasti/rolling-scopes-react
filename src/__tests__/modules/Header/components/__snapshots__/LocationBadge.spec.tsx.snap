// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationBadge location badge element should be rendered and match snapshot 1`] = `
<div>
  <div
    class="sc-bczRLJ oDGbA"
  >
    <p
      class="sc-gsnTZi iueXfi"
    >
      Location: 
      <span
        class="sc-dkzDqf buRVPe"
      >
        /test
      </span>
    </p>
    <p
      class="sc-gsnTZi iueXfi"
    >
      Component: 
      <span
        class="sc-dkzDqf buRVPe"
      >
        Test
      </span>
    </p>
  </div>
</div>
`;
